src/deluge/gui/views/session_view.cpp

```diff
@@ -4609,3 +4638,157 @@ Output* SessionView::getOutputFromPad(int32_t x, int32_t y) {
 	}
 	return nullptr;
 }
+
+void SessionView::gridCoordsFromClip(Clip* clip, uint32_t& x, uint32_t& y) {
+	Output* track = clip->output;
+	auto maxTrack = gridTrackCount();
+	int32_t trackIndex = gridTrackIndexFromTrack(track, maxTrack);
+	x = gridXFromTrack(trackIndex);
+	y = gridYFromSection(clip->section);
+}
+
+// stop pulsing selected clip
+// called when exiting session view
+void SessionView::gridStopSelectedClipPulsing() {
+	uiTimerManager.unsetTimer(TimerName::SELECTED_CLIP_PULSE);
+	gridSelectedClipPulsing = false;
+	selectedClipForPulsing = nullptr;
+	gridResetSelectedClipPulseProgress();
+}
+
+// reset blend position for pulse
+void SessionView::gridResetSelectedClipPulseProgress() {
+	progress = kMinProgress;
+	blendDirection = 1;
+	gridSelectedClipRenderedColour = RGB::monochrome(0);
+}
+
+// render pulse for selected clip
+void SessionView::gridSelectClipForPulsing(Clip* clip) {
+	// save the selected clip so that we don't override the pulsed colour when re-rendering grid
+	gridResetSelectedClipPulseProgress();
+	selectedClipForPulsing = clip;
+}
+
+// check if we should stop pulsing
+bool SessionView::gridCheckForPulseStop() {
+	// don't render pulse if we're in another view
+	if (getCurrentUI() != this) {
+		return true;
+	}
+	// don't render pulse if we're in row view
+	else if (currentSong->sessionLayout == SessionLayoutType::SessionLayoutTypeRows) {
+		return true;
+	}
+	// don't render pulse if we're transiting from session view to another view
+	else if (currentUIMode == UI_MODE_EXPLODE_ANIMATION) {
+		return true;
+	}
+	return false;
+}
+
+static const uint32_t pulseUIModes[] = {UI_MODE_CLIP_PRESSED_IN_SONG_VIEW, UI_MODE_CREATING_CLIP, 0};
+
+void SessionView::gridPulseSelectedClip() {
+	if (gridSelectedClipPulsing) {
+		// check if we should stop pulsing
+		if (gridCheckForPulseStop()) {
+			return gridStopSelectedClipPulsing();
```

Comment:
The return should be on its own line, otherwise this will silently cause bugs if gridStopSelectedClipPulsing ever has a return value added (e.g. maybe an error code)